crates/pico/src/memo_ref.rs

```diff
@@ -27,12 +80,23 @@ impl<T> PartialEq for MemoRef<T> {
 
 impl<T> Eq for MemoRef<T> {}
 
+#[allow(clippy::unnecessary_cast)]
+impl<T> Hash for MemoRef<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let data_ptr = self.db as *const dyn DatabaseDyn as *const ();
+        data_ptr.hash(state);
+        self.derived_node_id.hash(state);
+    }
+}
+
 #[allow(clippy::unnecessary_cast)]
 impl<T: 'static + Clone> MemoRef<T> {
     pub fn new(db: &dyn DatabaseDyn, derived_node_id: DerivedNodeId) -> Self {
         Self {
             db: db as *const _ as *const dyn DatabaseDyn,
             derived_node_id,
+            projectors: [step_identity; MAX_PROJECTOR_STEPS],
+            projectors_len: 0,
```

Comment:
Coming back to this... if the typical thing is `len == 0` (is it?) then, perhaps we can take inspiration from how iterators work: https://doc.rust-lang.org/src/core/iter/adapters/map.rs.html#63 they are structs that wrap the parent, and have `impl MemoRef<Item = T>` everywhere? Then, we need only `Map` and `FlatMap` structs that wrap the underlying memo ref.

Is that doable? Worse than this solution? I don't know. I'm just pattern-matching on how stuff is done elsewhere in Rust